File: persistence/src/main/java/io/sparkled/persistence/stage/impl/StagePersistenceServiceImpl.java
package io.sparkled.persistence.stage.impl;

import io.sparkled.model.entity.Stage;
import io.sparkled.model.entity.StageProp;
import io.sparkled.persistence.QueryFactory;
import io.sparkled.persistence.stage.StagePersistenceService;
import io.sparkled.persistence.stage.impl.query.*;

import javax.inject.Inject;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

public class StagePersistenceServiceImpl implements StagePersistenceService {

    private QueryFactory queryFactory;

    @Inject
    public StagePersistenceServiceImpl(QueryFactory queryFactory) {
        this.queryFactory = queryFactory;
    }

    @Override
    public Stage createStage(Stage stage) {
        return new SaveStageQuery(stage).perform(queryFactory);
    }

    @Override
    public List<Stage> getAllStages() {
        return new GetAllStagesQuery().perform(queryFactory);
    }

    @Override
    public Optional<Stage> getStageById(int stageId) {
        return new GetStageByIdQuery(stageId).perform(queryFactory);
    }

    @Override
    public List<StageProp> getStagePropsByStageId(Integer stageId) {
        return new GetStagePropsByStageIdQuery(stageId).perform(queryFactory);
    }

    @Override
    public Optional<StageProp> getStagePropByUuid(Integer stageId, UUID uuid) {
        return new GetStagePropByUuidQuery(stageId, uuid).perform(queryFactory);
    }

    @Override
    public void saveStage(Stage stage, List<StageProp> stageProps) {
        stage = new SaveStageQuery(stage).perform(queryFactory);
        new SaveStagePropsQuery(stage, stageProps).perform(queryFactory);
    }

    @Override
    public void deleteStage(int stageId) {
        new DeleteStageQuery(stageId).perform(queryFactory);
    }
}

File: persistence/src/main/kotlin/io/sparkled/persistence/stage/impl/StagePersistenceServiceImpl.kt
package io.sparkled.persistence.stage.impl

import io.sparkled.model.entity.Stage
import io.sparkled.model.entity.StageProp
import io.sparkled.persistence.QueryFactory
import io.sparkled.persistence.stage.StagePersistenceService
import io.sparkled.persistence.stage.impl.query.DeleteStageQuery
import io.sparkled.persistence.stage.impl.query.GetAllStagesQuery
import io.sparkled.persistence.stage.impl.query.GetStageByIdQuery
import io.sparkled.persistence.stage.impl.query.GetStagePropByUuidQuery
import io.sparkled.persistence.stage.impl.query.GetStagePropsByStageIdQuery
import io.sparkled.persistence.stage.impl.query.SaveStagePropsQuery
import io.sparkled.persistence.stage.impl.query.SaveStageQuery
import java.util.UUID
import javax.inject.Inject

class StagePersistenceServiceImpl
@Inject constructor(private val queryFactory: QueryFactory) : StagePersistenceService {

    override fun createStage(stage: Stage): Stage {
        return SaveStageQuery(stage).perform(queryFactory)
    }

    override fun getAllStages(): List<Stage> {
        return GetAllStagesQuery().perform(queryFactory)
    }

    override fun getStageById(stageId: Int): Stage? {
        return GetStageByIdQuery(stageId).perform(queryFactory)
    }

    override fun getStagePropsByStageId(stageId: Int): List<StageProp> {
        return GetStagePropsByStageIdQuery(stageId).perform(queryFactory)
    }

    override fun getStagePropByUuid(stageId: Int, uuid: UUID): StageProp? {
        return GetStagePropByUuidQuery(stageId, uuid).perform(queryFactory)
    }

    override fun saveStage(stage: Stage, stageProps: List<StageProp>) {
        val savedStage = SaveStageQuery(stage).perform(queryFactory)
        SaveStagePropsQuery(savedStage, stageProps).perform(queryFactory)
    }

    override fun deleteStage(stageId: Int) {
        DeleteStageQuery(stageId).perform(queryFactory)
    }
}
